### PR TITLE
Support MinBy and MaxBy for non-float ordering

### PR DESCRIPTION
**Pull request by thirtiseven**
Closes https://github.com/NVIDIA/spark-rapids/issues/10968

This PR supports MinBy and MaxBy aggregation for non-float ordering, which is not required by the customer.
 
The main idea is to pack the ordering and value columns to a struct column and do a min/max aggregation on them. When the ordering column is not unique this pr will always return the minimal value while spark's result is non-deterministic.

So this pr is not using the previous argmin + remove nulls + gather method in https://github.com/rapidsai/cudf/pull/16163 and https://github.com/NVIDIA/spark-rapids/pull/11123. It does not have cudf dependency pr now.

Spark has a special NaN handling which is different with cuDF, so it will be not straightforward to implement float ordering min/max_by with current method. But it is a rare use case I think we don't need to support for now.

Many works are from @firestarman 

Ran a simple perf test for 3 times

```scala
// (with big data gen)
import org.apache.spark.sql.tests.datagen._
val dataTable = DBGen().addTable("data", "a byte, b long, c long", 500000000)
dataTable.toDF(spark).write.mode("OVERWRITE").parquet("PERF")

// spark-rapids
val df = spark.read.parquet("PERF")
spark.time(df.groupBy("a").agg(min_by(col("b"), col("c")), min_by(col("c"), col("b"))).show(300))
```
on 500000000 rows of (a: byte, b: long, c: long)

results:

```
gpu: 3,484 ms
cpu: 5,844 ms
speedup: 68%
```

Another round with a complex value and simple order:
```
val dataTable = DBGen().addTable("data", "a byte, b struct<string, array<long>>, c int", 20000000)
spark.time(df.groupBy("a").agg(min_by(col("b"), col("c"))).show(300, false))

gpu: 1,049 ms
#11123: 1,019 m s
cpu: 2,776 ms
speedup: 2.65x
```

### CODE REVIEW

File: integration_tests/src/main/python/data_gen.py [hunk 1]
```diff
@@ -541,6 +541,35 @@ def make_tuple():
     def contains_ts(self):
         return any(child[1].contains_ts() for child in self.children)
 
+class StructGen2(DataGen):
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:544]: Why do we need/want this?  You can do the same thing with a function.

```
def my_struct_gen(child1, child2, nullable=True, special_cases=[]):
  return StructGen([('a', child1),('b', child2),('c', child2)], nullable=nullable, special_cases=special_cases)
```

File: integration_tests/src/main/python/data_gen.py [hunk 2]
```diff
@@ -541,6 +541,35 @@ def make_tuple():
...
+    def start(self, rand):
+        self.child1.start(rand)
+        self.child2.start(rand)
+        def make_tuple():
+            tmp_child2_gen = self.child2.gen()
+            data = [self.child1.gen(), tmp_child2_gen, tmp_child2_gen]
+            return tuple(data)
+        self._start(rand, make_tuple)
+
+def three_col_df2(spark, a_gen, b_gen, length=2048, seed=None, num_slices=None):
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:569]: This is named really badly, and I don't see why we need it when all you need is to repeat b_gen twice

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 3]
```diff
@@ -2023,3 +2023,143 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+    input: Expression, 
+    mask: Expression) extends GpuExpression with ShimExpression {
+
+  override def dataType: DataType = input.dataType
+  override def nullable: Boolean = mask.nullable
+  override def children: Seq[Expression] = Seq(input, mask)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    val res = withResource(input.columnarEval(batch)) { inputColumn =>
+      withResource(mask.columnarEval(batch)) { maskColumn =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2037]: maskCloumn can be closed right away after getting the maskColumnNotNull column to reduce te peak memory usage during executing.
```
val maskColumnNotNull = withResource(mask.columnarEval(batch)) { maskColumn =>
  maskColumn.getBase.isNotNull
}
withResource(maskColumnNotNull) { _ =>
  inputColumn.getBase.copyWithBooleanColumnAsValidity(maskColumnNotNull)
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 4]
```diff
@@ -2023,3 +2023,143 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+abstract class CudfMaxMinByAggregate(
+    orderingType: DataType,
+    valueType: DataType) extends CudfAggregate {
+
+  protected val sortOrder: Int => cudf.OrderByArg
+
+  protected val reductionAggregation: ReductionAggregation
+
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
+    val orderCol = col.getChildColumnView(0).copyToColumnVector()
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2061]: This code is not the latest, missing the handle for the all-nulls case.
```
  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
    val orderColView = col.getChildColumnView(1)
    if (orderColView.getNullCount == orderColView.getRowCount) { // all nulls
      GpuScalar.from(null, dataType)
    } else {
      val tmpTable = withResource(orderColView.copyToColumnVector())(
        orderCol => new cudf.Table(orderCol, col)
      )
      val sorted = withResource(tmpTable) { _ =>
        // columns in table [order, original struct]
        tmpTable.orderBy(sortOrder(0))
      }
      withResource(sorted) { _ =>
        sorted.getColumn(1).reduce(ReductionAggregation.nth(0, NullPolicy.INCLUDE))
      }
    }
  }
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 5]
```diff
@@ -2023,3 +2023,143 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+    AttributeReference("ordering", orderingExpr.dataType)()
+
+  private lazy val bufferValue: AttributeReference =
+    AttributeReference("value", valueExpr.dataType)()
+
+  // Cudf allows only one column as input, so wrap value and ordering columns by
+  // a struct before just going into cuDF.
+  private def createStructExpression(order: Expression, value: Expression): Expression = 
+    GpuReplaceNullmask(
+    GpuCreateNamedStruct(Seq(
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2112]: NIT: incorrect indents

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: removed

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala (hunk 4 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: thanks, updated

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala (hunk 5 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: done

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala (hunk 3 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 6]
```diff
@@ -2269,6 +2269,64 @@ object GpuOverrides extends Logging {
...
+      ),
+      (maxBy, conf, p, r) => new AggExprMeta[MaxBy](maxBy, conf, p, r) {
+
+        override def convertToGpu(childExprs: Seq[Expression]): GpuExpression = {
+          // Only two children (value expression, ordering expression)
+          require(childExprs.length == 2)
+          GpuMaxBy(childExprs.head, childExprs.last)
+        }
+
+        // MinBy does not overflow, so it doesn't need the ANSI check
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2298]:
```suggestion
        // MaxBy does not overflow, so it doesn't need the ANSI check
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 7]
```diff
@@ -2023,3 +2023,136 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+    orderingType: DataType,
+    valueType: DataType) extends CudfAggregate {
+
+  protected val reductionAggregation: ReductionAggregation
+
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
+    if (col.getNullCount == col.getRowCount) { // all nulls
+      GpuScalar.from(null, dataType)
+    } else {
+      withResource(col.copyToColumnVector()) { colVec =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2063]: AFAIK, the input "col" is already a ColumnVector, why does it still need to call "copyToColumnVector" here ?

Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: Thanks, fixed, it doesn't.

**Comment by thirtiseven** [comment]
build

**Comment by firestarman** [comment]
Look good to me

**Comment by thirtiseven** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 8]
```diff
@@ -2023,3 +2023,134 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+  override lazy val groupByAggregate: GroupByAggregation = GroupByAggregation.min()
+  override lazy val reductionAggregation: ReductionAggregation = ReductionAggregation.min()
+}
+
+abstract class GpuMaxMinByBase(valueExpr: Expression, orderingExpr: Expression)
+  extends GpuAggregateFunction with Serializable {
+
+  protected val cudfMaxMinByAggregate: CudfAggregate
+
+    private lazy val bufferOrdering: AttributeReference =
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2093]: nit: indentation appears to be off.

**Review by revans2 — COMMENTED** [review]
I am also concerned with the performance numbers. I don't know how to reproduce those numbers because df is not included. A speedup of 36% over the CPU is not much, and I am very concerned that if the value of the aggregation is long and complicated while the ordering key is not we will lose to the CPU because the CPU just has to do ordering for a very simple operation.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 9]
```diff
@@ -2023,3 +2023,134 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
...
+
+case class GpuReplaceNullmask(
+    input: Expression, 
+    mask: Expression) extends GpuExpression with ShimExpression {
+
+  override def dataType: DataType = input.dataType
+  override def nullable: Boolean = mask.nullable
+  override def children: Seq[Expression] = Seq(input, mask)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2035]: https://github.com/NVIDIA/spark-rapids/issues/11397

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 10]
```diff
@@ -1339,6 +1339,43 @@ def test_generic_reductions(data_gen):
             'count(1)'),
         conf=local_conf)
 
+# min_by and max_by are supported for pyspark since 3.3.0 so tested with sql
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen', all_basic_gens + struct_gens_sample_with_decimal128 + array_gens_sample, ids=idfn)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/hash_aggregate_test.py]: Added

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala:2035]: Thanks, updated to `ifElse`

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala (hunk 8 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: done

File: tools/generated_files/311/operatorsScore.csv [hunk 11]
```diff
@@ -0,0 +1,272 @@
+CPUOperator,Score
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply tools/generated_files/311/operatorsScore.csv]: removed

File: tools/generated_files/312/operatorsScore.csv [hunk 12]
```diff
@@ -0,0 +1,272 @@
+CPUOperator,Score
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply tools/generated_files/312/operatorsScore.csv]: removed

File: tools/generated_files/312/supportedExprs.csv [hunk 13]
```diff
@@ -0,0 +1,746 @@
+Expression,Supported,SQL Func,Notes,Context,Params,BOOLEAN,BYTE,SHORT,INT,LONG,FLOAT,DOUBLE,DATE,TIMESTAMP,STRING,DECIMAL,NULL,BINARY,CALENDAR,ARRAY,MAP,STRUCT,UDT
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply tools/generated_files/312/supportedExprs.csv]: removed

File: tools/generated_files/313/supportedExprs.csv [hunk 14]
```diff
@@ -0,0 +1,746 @@
+Expression,Supported,SQL Func,Notes,Context,Params,BOOLEAN,BYTE,SHORT,INT,LONG,FLOAT,DOUBLE,DATE,TIMESTAMP,STRING,DECIMAL,NULL,BINARY,CALENDAR,ARRAY,MAP,STRUCT,UDT
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply tools/generated_files/313/supportedExprs.csv]: removed

**Comment by thirtiseven** [comment]
> I am also concerned with the performance numbers. I don't know how to reproduce those numbers because df is not included. A speedup of 36% over the CPU is not much, and I am very concerned that if the value of the aggregation is long and complicated while the ordering key is not we will lose to the CPU because the CPU just has to do ordering for a very simple operation.

Thanks for the review! Added data generation part in pr description, and another round test of `"a byte, b struct<string, array<long>>, c int"`, it got 2.65x speedup.

I think in most use cases the ordering key is unique, so the compare function will almost always stop early before getting into the complicated value part, so it won't be slower.

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
@thirtiseven could you please file a follow on issue to look at ways to improve the performance of these operators. I did my own benchmarking and this is far from an ideal solution. CUDF does not deal well with min/max on a struct of two values. I ran the following benchmarks.

I have 4 different data sets and 4 separate queries on those data sets.  All of the tests were run on a 32 core thread ripper with spark configured to use 32 cores. The GPU is an a6000, but limited to 16 GiB of GPU memory.

Dataset 1: (100 group by keys)
```
spark.range(10000000000L).selectExpr("id div 100000000 as a", "id div 1000000 as b", "id as c").createOrReplaceTempView("tbl")
```

Dataset 2: (100 group by keys, but made into structs before agg) The 2048 partitions is to eliminate spilling on the sort for the CPU.  I also ran one with the GPU at 64 partitions to make it more comparable to other CPU jobs.
```
spark.range(1, 10000000000L, 1 , 2048).selectExpr("id div 100000000 as a", "named_struct('f', id div 1000000, 'o', id) as b", "named_struct('o', id, 'f', id div 1000000) as c").createOrReplaceTempView("tbl")
```

Dataset 3: (1 million group by keys)
```
spark.range(10000000000L).selectExpr("id div 10000 as a", "id div 100 as b", "id as c").createOrReplaceTempView("tbl")
```

Dataset 4: (1 million group by keys, but made into structs before agg) The 2048 partitions is to eliminate spilling on the sort for the CPU. I also ran one with the GPU at 64 partitions to make it more comparable to other CPU jobs.
```
spark.range(1, 10000000000L, 1 , 2048).selectExpr("id div 1000 as a", "named_struct('f', id div 100, 'o', id) as b", "named_struct('o', id, 'f', id div 100) as c").createOrReplaceTempView("tbl")
```

Query 1: (group by min_by and max_by)
```
spark.time(spark.sql("SELECT a, min_by(c,b), max_by(c,b), min_by(b, c), max_by(b, c) FROM tbl GROUP BY a ORDER BY a").show())
```

Query 2: (group by min and max)
```
spark.time(spark.sql("SELECT a, min(c), max(c), min(b), max(b) FROM tbl GROUP BY a ORDER BY a").show())
```

Query 3: (min_by and max_by reduction)
```
spark.time(spark.sql("SELECT min_by(c,b), max_by(c,b), min_by(b, c), max_by(b, c) FROM tbl").show())
```

Query 4: (min and max reduction)
```
spark.time(spark.sql("SELECT min(c), max(c), min(b), max(b) FROM tbl").show())
```


hardware | data set | Q1 | Q2 | Q3 | Q4
-- | -- | -- | -- | -- | --
GPU | DS1 | 42,988 | 7,508 | 4,689 | 952
CPU | DS1 | 11,602 | 6,549 | 2,822 | 1,540
GPU | DS2(2048) | 56,256 | 44,122 | 23,232 | 7,053
CPU | DS2(2048) | 482,006 | 272,889 | 418,851 | 222,792
GPU | DS2(64) | 50,855 | 41,485 | 9,438 | 4,074
GPU | DS3 | 42,686 | 3,698 | 4,518 | 953
CPU | DS3 | 10,837 | 7,073 | 2,645 | 1,540
GPU | DS4(2048) | 57,188 | 45,757 | 20,856 | 6,928
CPU | DS4(2048) | 492,970 | 289,821 | 418,500 | 221,631
GPU | DS4(64) | 52,555 | 42,694 | 9,199 | 3,896

A few things to note from the results.

   1. A min/max on a `long` (DS1 - Q2) is about 5.7 times faster than a `struct(long, long)` (DS1 - Q1) (because in CUDF we are forced to fall back to a sort aggregate for the struct operation.
   2. The CPU saw less than a 2x speedup for the same change.
   3. For DS2(2048) and DS2(64)  the GPU saw very little speedup because now both are sorted.
   4. For DS2(2048) the CPU saw a very similar speedup of a little less that 2x.
   5. We know that GPU has some issue with a small number of group by keys in aggregations compared to reductions. The CPU shows no such problems.

In general I think if we went with the `argmin` and then a gather for many fixed width types we would see a huge performance improvement. If the they are not fixed width, then we would not likely see much performance improvement, but we might see a very small amount, just because there would be less data to look at.

**Comment by thirtiseven** [comment]
@revans2 Thanks for the testing and analysis! filed https://github.com/NVIDIA/spark-rapids/issues/11412

**Pull request merged**